### PR TITLE
Add `DataPayload` to `Transaction` for data storage. 

### PR DESCRIPTION
**Pull request by MichaelKim20**
I add a `DataPayload` to `Transaction` for data storage. 

Relates to #1214

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
This should fix your errors:

```diff
diff --git a/source/agora/consensus/data/DataPayload.d b/source/agora/consensus/data/DataPayload.d
index 5b81a6da2..8e2d7141c 100644
--- a/source/agora/consensus/data/DataPayload.d
+++ b/source/agora/consensus/data/DataPayload.d
@@ -35,7 +35,7 @@ public struct DataPayload
 
     ***************************************************************************/
 
-    public this (const(ubyte)[] bin) pure nothrow @safe
+    public this (inout(ubyte)[] bin) inout pure nothrow @safe
     {
         this.data = bin;
     }
```

File: source/agora/consensus/data/DataPayload.d [hunk 1]
```diff
@@ -35,7 +35,7 @@ public struct DataPayload
 
     ***************************************************************************/
 
-    public this (const(ubyte)[] bin) pure nothrow @safe
+    public this (inout(ubyte)[] bin) inout pure nothrow @safe
```
Comment (MichaelKim20) [inline source/agora/consensus/data/DataPayload.d:38]: @AndrejMitrovic  helped me.
https://github.com/bpfkorea/agora/pull/1368#issuecomment-727737956

File: source/agora/node/Fee.d [hunk 2]
```diff
@@ -155,3 +160,95 @@ unittest
...
+            tx = `Transaction`
+            sum_unspent = Sum of unspent amounts
+
+        Return:
+            `null` if the transaction is valid, a string explaining the reason it
+            is invalid otherwise.
+
+    ***************************************************************************/
+
+    public string check (in Transaction tx, in Amount sum_unspent) nothrow @safe
```
Comment (MichaelKim20) [inline source/agora/node/Fee.d]: https://github.com/bpfkorea/agora/pull/1321#issuecomment-727713270
The above link was proposed by @AndrejMitrovic  in the previous PR.
I tried to make a simple delegate that only checks fees according to @AndrejMitrovic 's suggestion.
But I did this because I needed more information (Commons Budget Address, ...).

If this is to be changed to a simple delegate to check fee, then ConsensusParam must be delivered to `agora/consensus/validation/Transaction.d`
I did this to reduce dependencies, but please let me know if you have any other opinions.

**Comment by MichaelKim20** [comment]
I just added a module and a test code to check the data payload.
Ready for review

File: source/agora/common/Config.d [hunk 3]
```diff
@@ -168,6 +168,12 @@ public struct NodeConfig
     /// is going to connect to, for example: http://0.0.0.0:8008
     /// It can also be set to 0 do disable listening
     public ushort stats_listening_port;
+
+    /// The maximum size of data payload
+    public uint max_size_payload;
```
Comment (AndrejMitrovic) [inline source/agora/common/Config.d]: Can we rename this to `tx_payload_max_size`?

File: source/agora/common/Config.d [hunk 4]
```diff
@@ -168,6 +168,12 @@ public struct NodeConfig
...
+
+    /// The factor to calculate for the fee of data payload
+    public uint payload_fee_factor;
```
Comment (AndrejMitrovic) [inline source/agora/common/Config.d]: Can we rename this to `tx_payload_fee_factor`?

File: source/agora/common/Config.d (hunk 3 above)
Comment (AndrejMitrovic) [reply source/agora/common/Config.d]: It should have a default value, no?

File: source/agora/common/Config.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/agora/common/Config.d]: Also: It should have a default value, no?

File: tests/system/node/0/config.yaml [hunk 5]
```diff
@@ -15,6 +15,10 @@ node:
   # Path to the data directory (if the path doesn't exist it will be created)
   data_dir: .cache
 
+  commons_budget_address: GCOMMONBGUXXP4RFCYGEF74JDJVPUW2GUENGTKKJECDNO6AGO32CUWGU
```
Comment (AndrejMitrovic) [inline tests/system/node/0/config.yaml]: This seems to be an unrelated change. Please move it to another commit, or remove it entirely.

File: tests/system/node/2/config.yaml [hunk 6]
```diff
@@ -14,6 +14,10 @@ node:
   # Path to the data directory (if the path doesn't exist it will be created)
   data_dir: .cache
 
+  commons_budget_address: GCOMMONBGUXXP4RFCYGEF74JDJVPUW2GUENGTKKJECDNO6AGO32CUWGU
```
Comment (AndrejMitrovic) [inline tests/system/node/2/config.yaml]: This seems to be an unrelated change. Please move it to another commit, or remove it entirely.

File: tests/system/node/3/config.yaml [hunk 7]
```diff
@@ -15,6 +15,10 @@ node:
   # Path to the data directory (if the path doesn't exist it will be created)
   data_dir: .cache
 
+  commons_budget_address: GCOMMONBGUXXP4RFCYGEF74JDJVPUW2GUENGTKKJECDNO6AGO32CUWGU
```
Comment (AndrejMitrovic) [inline tests/system/node/3/config.yaml]: This seems to be an unrelated change. Please move it to another commit, or remove it entirely.

File: tests/system/node/4/config.yaml [hunk 8]
```diff
@@ -15,6 +15,10 @@ node:
   # Path to the data directory (if the path doesn't exist it will be created)
   data_dir: .cache
 
+  commons_budget_address: GCOMMONBGUXXP4RFCYGEF74JDJVPUW2GUENGTKKJECDNO6AGO32CUWGU
```
Comment (AndrejMitrovic) [inline tests/system/node/4/config.yaml]: This seems to be an unrelated change. Please move it to another commit, or remove it entirely.

File: tests/system/node/5/config.yaml [hunk 9]
```diff
@@ -15,6 +15,10 @@ node:
   # Path to the data directory (if the path doesn't exist it will be created)
   data_dir: .cache
 
+  commons_budget_address: GCOMMONBGUXXP4RFCYGEF74JDJVPUW2GUENGTKKJECDNO6AGO32CUWGU
```
Comment (AndrejMitrovic) [inline tests/system/node/5/config.yaml]: This seems to be an unrelated change. Please move it to another commit, or remove it entirely.

File: tests/system/node/6/config.yaml [hunk 10]
```diff
@@ -15,6 +15,10 @@ node:
   # Path to the data directory (if the path doesn't exist it will be created)
   data_dir: .cache
 
+  commons_budget_address: GCOMMONBGUXXP4RFCYGEF74JDJVPUW2GUENGTKKJECDNO6AGO32CUWGU
```
Comment (AndrejMitrovic) [inline tests/system/node/6/config.yaml]: This seems to be an unrelated change. Please move it to another commit, or remove it entirely.

File: tests/system/node/7/config.yaml [hunk 11]
```diff
@@ -15,6 +15,10 @@ node:
   # Path to the data directory (if the path doesn't exist it will be created)
   data_dir: .cache
 
+  commons_budget_address: GCOMMONBGUXXP4RFCYGEF74JDJVPUW2GUENGTKKJECDNO6AGO32CUWGU
```
Comment (AndrejMitrovic) [inline tests/system/node/7/config.yaml]: This seems to be an unrelated change. Please move it to another commit, or remove it entirely.

File: tests/system/node/0/config.yaml [hunk 12]
```diff
@@ -15,6 +15,10 @@ node:
...
+  max_size_payload: 1024
```
Comment (AndrejMitrovic) [inline tests/system/node/0/config.yaml]: If you set the default value in `Config.d`, then you don't have to manually specify the values here.

File: source/agora/consensus/data/Params.d [hunk 13]
```diff
@@ -54,6 +54,12 @@ public immutable class ConsensusParams
     /// The address of commons budget
     public PublicKey CommonsBudgetAddress;
 
+    /// The maximum size of the payload
+    public uint MaxSizePayload;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/Params.d]: Can we rename to `TxPayloadMaxSize`?

File: source/agora/consensus/data/Params.d [hunk 14]
```diff
@@ -54,6 +54,12 @@ public immutable class ConsensusParams
...
+
+    /// The factor to calculate for the fee of the payload
+    public uint PayloadFeeFactor;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/Params.d]: Can we rename to `TxPayloadFeeFactor`?

File: source/agora/consensus/state/UTXOSet.d [hunk 15]
```diff
@@ -71,6 +71,11 @@ abstract class UTXOCache
             unlock_height = Height(height + 2016);
         }
 
+        // A comment with the introduction of a new transaction type (TxType.Payload)
+        // If a transaction type is `TxType.Payload`, and the input is `TxType.Freeze`,
+        // the transaction is already excluded at `validation`.
+        // No conditions have been added for `TxType.Payload` here.
+
```
Comment (AndrejMitrovic) [inline source/agora/consensus/state/UTXOSet.d]: This comment belongs in the commit message itself, not here.

File: source/agora/consensus/validation/Block.d [hunk 16]
```diff
@@ -179,6 +181,8 @@ public string isGenesisBlockInvalidReason (const ref Block block) nothrow @safe
     UTXO[Hash] utxo_set;
     foreach (const ref tx; block.txs)
     {
+        // Data storage type(TxType.Payload) is currently not required for Genesis Block,
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]: It would be better to explicitly disallow payload transactions in the genesis block.

File: source/agora/consensus/data/Transaction.d [hunk 17]
```diff
@@ -31,6 +32,7 @@ public enum TxType : ubyte
 {
     Payment,
     Freeze,
+    Payload
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/Transaction.d]: Why are you introducing a new transaction type?

If the transaction's payload array is not empty, then it's a payload transaction.

If you remove this special `Payload` TxType then you can remove all this special case code.

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
Please remove `TxType.Payload`. It's unnecessary and makes the code changes large and complexity hard.

File: source/agora/consensus/data/Transaction.d (hunk 17 above)
Comment (MichaelKim20) [reply source/agora/consensus/data/Transaction.d]: Thank you for your good suggestion. 
I have just completed the modification according to your suggestion.

File: source/agora/common/Config.d (hunk 4 above)
Comment (MichaelKim20) [reply source/agora/common/Config.d]: This is a good suggestion. I'll fix it.

File: source/agora/consensus/validation/Block.d [hunk 18]
```diff
@@ -188,6 +190,9 @@ public string isGenesisBlockInvalidReason (const ref Block block) nothrow @safe
         if (tx.outputs.length == 0)
             return "GenesisBlock: No output(s) in the transaction";
 
+        if (tx.payload.data.length != 0)
+            return "GenesisBlock: The data payload cannot be stored";
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d:195]: This line is not tested according to https://codecov.io/gh/bpfkorea/agora/pull/1368/src/source/agora/consensus/validation/Block.d.

Please make sure to always add a test when adding new consensus rules. :)

File: source/agora/consensus/validation/Transaction.d [hunk 19]
```diff
@@ -122,6 +128,13 @@ public string isInvalidReason (
         // current limitation: if any UTXO is frozen, they all must be frozen
         if ((count_freeze > 0) && (count_freeze != tx.inputs.length))
             return "Transaction: Rejected combined inputs (freeze & payment)";
+
+        // current limitation: frozen UTXO cannot be used for data payload.
+        if ((tx.payload.data.length != 0) && (count_freeze > 0))
+            return "Transaction: Frozen UTXO cannot be used for data payload";
+
+        auto res = checkPayload(tx, sum_unspent);
+        if (res !is null) return res;
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Transaction.d]:
```D
if (auto error = checkPayload(tx, sum_unspent))
    return error;
```

File: source/agora/consensus/validation/Transaction.d [hunk 20]
```diff
@@ -95,6 +98,9 @@ public string isInvalidReason (
 
         if (sum_unspent.integral() < Amount.MinFreezeAmount.integral())
             return "Transaction: available when the amount is at least 40,000 BOA";
+
+        auto res = checkPayload(tx, sum_unspent);
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Transaction.d]: I think here we should disallow a data payload if the transaction is a freeze transaction. So similar to how you did it below:

```D
if (tx.payload.data.length != 0)
    return "Transaction: Frozen UTXO cannot be used for data payload";
```

Also, could you move this check more up? If we check this before checking the `sum_unspent`, it will be more efficient.

File: source/agora/consensus/validation/Transaction.d [hunk 21]
```diff
@@ -122,6 +128,13 @@ public string isInvalidReason (
         // current limitation: if any UTXO is frozen, they all must be frozen
         if ((count_freeze > 0) && (count_freeze != tx.inputs.length))
             return "Transaction: Rejected combined inputs (freeze & payment)";
+
+        // current limitation: frozen UTXO cannot be used for data payload.
+        if ((tx.payload.data.length != 0) && (count_freeze > 0))
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Transaction.d]: No this check is wrong. If we are spending a melted transaction we can do whatever we want with the payment transaction.

The rules should be:

Payment: Can have data payload
Payment which spends a melted freeze: Can have a data payload
Frozen: Cannot have data payload

So your Freeze and Payment checks are inverted in this PR.

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]

File: source/agora/consensus/validation/Transaction.d (hunk 20 above)
Comment (MichaelKim20) [reply source/agora/consensus/validation/Transaction.d]: I think that's good, too.

File: source/agora/consensus/validation/Transaction.d (hunk 21 above)
Comment (MichaelKim20) [reply source/agora/consensus/validation/Transaction.d]: Okay Good

Comment (MichaelKim20) [reply source/agora/consensus/validation/Transaction.d]: Thank you for review, I fixed all

File: source/agora/node/Fee.d (hunk 2 above)
Comment (AndrejMitrovic) [reply source/agora/node/Fee.d]: Ah @MichaelKim20, please remove the `sum_unspent` parameter. I was mistaken that we need it. The fee is in the output for the commons budget, just like you set it here.

File: source/agora/node/Fee.d [hunk 22]
```diff
@@ -155,3 +160,95 @@ unittest
...
+    {
+        if (tx.payload.data.length == 0)
+            return null;
+
+        if (tx.payload.data.length > this.TxPayloadMaxSize)
+            return "Transaction: The size of the data payload is too large";
+
+        uint count_commons_budge = 0;
```
Comment (AndrejMitrovic) [inline source/agora/node/Fee.d]: Typo: `count_commons_budget`

**Review by AndrejMitrovic — COMMENTED** [review]
Just two comments, otherwise looks good.

**Review by AndrejMitrovic — APPROVED** [review]

**Review by TrustHenry — APPROVED** [review]
Nice! 
Due to the addition of `payload`, Stoa also needs to be applied.
Please make PR with Stoa before merging.

File: source/agora/consensus/validation/Block.d [hunk 23]
```diff
@@ -18,11 +18,13 @@ import agora.common.Hash;
 import agora.common.Set;
 import agora.common.Types;
 import agora.consensus.data.Block;
+import agora.consensus.data.DataPayload;
 import agora.consensus.data.Enrollment;
 import agora.consensus.data.Transaction;
 import agora.consensus.state.UTXOSet;
 import VEn = agora.consensus.validation.Enrollment;
 import VTx = agora.consensus.validation.Transaction;
+import agora.node.Fee;
```
Comment (Geod24) [inline source/agora/consensus/validation/Block.d:27]: I've said it many times: `consensus` should never import `node`
